File: 0503_next_greater_element_ii/Solution.swift
// LeetCode 0503 - Next Greater Element II
// https://leetcode.com/problems/next-greater-element-ii/

class Solution {
    func nextGreaterElements(_ nums: [Int]) -> [Int] {
        let length = nums.count
        var result = Array(repeating: -1, count: length)
        var stack: [Int] = []

        for index in 0..<(length * 2) {
            while !stack.isEmpty && nums[stack[stack.count - 1]] < nums[index % length] {
                result[stack.removeLast()] = nums[index % length]
            }
            if index < length {
                stack.append(index)
            }
        }
        return result
    }
}

File: 0503_next_greater_element_ii/Solution.kt
// LeetCode 0503 - Next Greater Element II
// https://leetcode.com/problems/next-greater-element-ii/

class Solution {
    fun nextGreaterElements(nums: IntArray): IntArray {
        val length = nums.size
        val result = IntArray(length) { -1 }
        val stack = ArrayDeque<Int>()
        for (index in 0 until length * 2) {
            val value = nums[index % length]
            while (stack.isNotEmpty() && nums[stack.last()] < value) {
                result[stack.removeLast()] = value
            }
            if (index < length) {
                stack.addLast(index)
            }
        }
        return result
    }
}
